src/preact/parse-props.js

```diff
@@ -0,0 +1,359 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '#preact';
+import {Loading} from '#core/loading-instructions';
+import {Slot, createSlot} from './slot';
+import {dashToCamelCase} from '#core/types/string';
+import {devAssert} from '#core/assert';
+import {getDate} from '#core/types/date';
+import {matches, realChildNodes} from '#core/dom/query';
+import {parseBooleanAttribute} from '#core/dom';
+import {sequentialIdGenerator} from '#core/math/id-generator';
+
+/**
+ * The following combinations are allowed.
+ * - `attr`, (optionally) `type`, and (optionally) `media` can be specified when
+ *   an attribute maps to a component prop 1:1.
+ * - `attrs` and `parseAttrs` can be specified when multiple attributes map
+ *   to a single prop.
+ * - `attrPrefix` can be specified when multiple attributes with the same prefix
+ *   map to a single prop object. The prefix cannot equal the attribute name.
+ * - `selector` can be specified for children of a certain shape and structure
+ *   according to ChildDef.
+ * - `passthrough` can be specified to slot children using a single
+ *   `<slot>` element for all children. This is in contrast to selector mode,
+ *   which creates a new named `<slot>` for every selector.
+ * - `passthroughNonEmpty` is similar to passthrough mode except that when there
+ *   are no children elements, the returned value will be null instead of the
+ *   unnamed `<slot>`. This allows the Preact environment to have conditional
+ *   behavior depending on whether or not there are children.
+ *
+ * @typedef {{
+ *   attr: (string|undefined),
+ *   type: (string|undefined),
+ *   attrPrefix: (string|undefined),
+ *   attrs: (!Array<string>|undefined),
+ *   parseAttrs: ((function(!Element):*)|undefined),
+ *   media: (boolean|undefined),
+ *   default: *,
+ * }|string}
+ */
+export let AmpElementPropDef;
+
+/**
+ * @typedef {{
+ *   name: string,
+ *   selector: string,
+ *   single: (boolean|undefined),
+ *   clone: (boolean|undefined),
+ *   props: (!JsonObject|undefined),
+ * }}
+ */
+let ChildDef;
+
+/** @const {string} */
+const RENDERED_ATTR = 'i-amphtml-rendered';
+
+/**
+ * The same as `applyFillContent`, but inside the shadow.
+ * @const {!Object}
+ */
+const SIZE_DEFINED_STYLE = {
+  'position': 'absolute',
+  'top': '0',
+  'left': '0',
+  'width': '100%',
+  'height': '100%',
+};
+
+/**
+ * This is an internal property that marks light DOM nodes that were rendered
+ * by AMP/Preact bridge and thus must be ignored by the mutation observer to
+ * avoid mutate->rerender->mutate loops.
+ */
+const RENDERED_PROP = '__AMP_RENDERED';
+
+const childIdGenerator = sequentialIdGenerator();
```

Comment:
How does `sequentialIdGenerator` work if being called for SSR and potentially also later on the client?

Reply:
Seems likely to me that it actually wouldn't work unless we use a `childIdGenerator` per-component